### PR TITLE
Ulcomp p1

### PR DESCRIPTION
**Pull request by munroesj52**
Allow the Used block List tree to combine adjacent blocks.
Closing #52

The current implementation of the used block tree is effectively a
linear list of allocated block ordered by size then address.
This creates two problems:
1) The list can get very large and may cause the anchor block heap to
run out of space.
2) The list becomes a singly linked list (not a tree) so searches,
inserts and removes slow down.

The change compacts the used block tree on block allocation by:
combining adjacent even/odd blocks of the same size into a single
block of the next larger power of 2 size. This is applied recursively
if larger blocks can be combined.

On deallocation the Used List is first searched for a node with match
size and address. If found this node is removed. Otherwise search
(p2FindCont) for a block, of larger size, containing the deallocated
block. If found, this block is split (p2split). The half that does not
contain the deallocated block, is reinserted in the list. The
containing half id checked for exact size and address match. If size
and address match, we are done and can insert the deallocated block
into the free list. Otherwise we continue spliting and containment
checking checking until we get down to the correct size and address.

	* src/sasanchr.h (regionFlags): Define bit fields for
	configuration flags.
	(rFlags): Replaces reserved1 field in SASAnchor_t struct.

	* src/sassim.cpp: add include <semaphore.h>.
	(SASS2Log2): New function.
	(p2FindCont, p2split, p2AddUsedCompact):
	New internal functions.
	(p2RemUsed [__SASDebugPrint__]): Additional debug printfs.
	(p2RemUsed): Add code the detect blocks contained within
	combined blocks and slit them as needed to isolate the
	orginal block for removal.
	(p2Dealloc [__SASDebugPrint__]): Additional debug printfs.
	(getSASUseListFlag, setSASLinearUseList, setSASCompactUseList):
	New function.
	(SASBlockAllocNoLock): If rFlags.compactUseList call
	p2AddUsedCompact else p2AddUsed.
	(SASJoinRegionByName): If creating and initializing a new
	SAS-store, set default using setSASCompactUseList.
	(SASMaxDepthAllocatedRegion, SASMaxDepthUseMem) New Functions.

	* src/sassim.h (SASS2Log2, getSASUseListFlag, 
	setSASLinearUseList, setSASCompactUseList,
	SASMaxDepthAllocatedRegion, SASMaxDepthUseMem): New externs

	* src/sasutil.c (sasutil_stat_cmd): New local vars; 
	cUsed, mUsed, cUsedReg, mUsedReg. Display Use List flag and
	node count and max depth for Use-List and Allocated-Region.
	(sasutil_detail_cmd): Display Use List flag,
	node count and max depth for Use-List and Allocated-Region.

	* src/tests/sassim_t.c [ARRAY_SIZE, MAX_ADDR_LIST,
	MAX_LINE_LEN]: New macros.
	(sassim_display_uselist_flag, sassim_display_summary,
	sassim_display_detail): New functions.
	[TEST_BLOCKS]: New macros.
	(blocks_s, blocks_m, blocks_l): New static arrays.
	(sassim_alloc_uselist): New unit test function.
	(main): Call sassim_alloc_uselist.

### CODE REVIEW

**Comment by munroesj52** [comment]
No reviews doe comments in 5 months. So unless someone speaks up, I will merge this as is.

**Comment by tuliom** [comment]
> No reviews or comments in 5 months. So unless someone speaks up, I will merge this as is.

I'm sorry, but I don't think I'll be able to review this PR soon.
I agree you've given more than enough time to get this merged.

File: src/sassim.cpp [hunk 1]
```diff
@@ -207,14 +214,150 @@ p2AddUsed (uLongTreeNode ** root, unsigned long size, void *loc)
...
+      toff = offset & ~((1L << (tsize + 12)) - 1);
+      keys = nodeToLong (toff, tsize);
+      nn = (*root)->searchEqualOrNextNode (root, keys);
+      if (nn)
+	{
+	  n = *nn;
+	  if (n)
+	    {
+	      k = n->getKey ();
+	      if ((k == keys) /*&& (l == (unsigned long) loc)*/)
```
Comment (rzinsly) [inline src/sassim.cpp:247]: Comment removed.

File: src/sassim.cpp [hunk 2]
```diff
@@ -207,14 +214,150 @@ p2AddUsed (uLongTreeNode ** root, unsigned long size, void *loc)
...
+	{
+	  sas_printf ("!SASFindCont integrity check failed\n");
+	  sas_printf ("  returning %p size %d\n", (void *) offset, lsize);
+	  sas_printf ("  trying    %p size %d\n", (void *) toff, tsize);
+	  if (n)
+	    {
+	      k = n->getKey ();
+	      toff = longToOffset (k);
+	      tsize = longToSize (k);
+	  sas_printf ("  last find %p size %d\n", (void *) toff, tsize);
```
Comment (rzinsly) [inline src/sassim.cpp:274]: Indentation fixed.

File: src/sassim.cpp [hunk 3]
```diff
@@ -207,14 +214,150 @@ p2AddUsed (uLongTreeNode ** root, unsigned long size, void *loc)
...
+  sas_printf ("p2split (%p, %u by %u)\n", (void*) tloc, csize, asize);
+#endif
+  SASNearDealloc (cblock, sizeof(uLongTreeNode));
+  tsize = csize;
+  toff = coff;
+
+#ifdef __SASDebugPrint__
+  sas_printf ("Split (%u:%u) (%lx:%lx)> ", tsize, asize, toff, aoff);
+#endif
+  while ((tsize > asize) /* & (toff != aoff)*/)
```
Comment (rzinsly) [inline src/sassim.cpp:312]: Comment removed.

File: src/sassim.cpp [hunk 4]
```diff
@@ -228,22 +371,38 @@ p2RemUsed (uLongTreeNode ** root, unsigned long size, void *loc)
...
+      else
+	{
+	  sas_printf ("!SASRemUsed1 integrity check failed\n");
+	  sas_printf ("  returning %p size %p\n", loc, (void *) size);
+	}
     }
 }
 
+#undef __SASDebugPrint__
+
```
Comment (rzinsly) [inline src/sassim.cpp:405]: Removed.

File: src/sassim.cpp [hunk 5]
```diff
@@ -397,9 +567,16 @@ p2Dealloc (uLongTreeNode ** root, unsigned long size, void *loc)
   n->init (keys, val);
   (*root)->insertNode (root, n);
 #ifdef __SASDebugPrint__
-  sas_printf ("\n");
+  sas_printf ("\n -->%p,%p\n", (void *) keys, (void *) val);
 #endif
 }
+#undef __SASDebugPrint__
```
Comment (rzinsly) [inline src/sassim.cpp:573]: Removed.

File: src/sassim.h [hunk 6]
```diff
@@ -151,6 +154,52 @@ extern __C__ unsigned long getMemHigh();
 extern __C__ void
 setSASmemrange (unsigned long low, unsigned long high);
 
+/** \brief Get UseList tracking mode flag
+*
+*   The in-use blocks can tracked as a linear list with each block
```
Comment (rzinsly) [inline src/sassim.h:159]: "can **be** tracked"

File: src/sassim.h [hunk 7]
```diff
@@ -238,6 +287,29 @@ extern __C__ void *SASBlockAlloc (unsigned long blockSize);
 */
 extern __C__ void SASBlockDealloc (void *blockAddr, unsigned long blockSize);
 
+/** \brief Return the maximum depth for the tree tracking allocated
+*   Region segments.
+*
+*   Returns the maximum depth of the tree of allocated allocated
```
Comment (rzinsly) [inline src/sassim.h:293]: Removed the repeated **allocated**.

File: src/tests/sassim_t.c [hunk 8]
```diff
@@ -547,6 +547,229 @@ sassim_space_test1 ()
...
+sassim_display_summary ()
+{
+  void *addrList[MAX_ADDR_LIST];
+  unsigned long sizeList[MAX_ADDR_LIST];
+  unsigned long tUsed, tFree, tUncom, tUsedReg, tFreeReg;
+  int count, maxd;
+  unsigned int anchorFree;
+  int i;
+
+//  printf ("Use List Flag =%d\n", getSASUseListFlag ());
```
Comment (rzinsly) [inline src/tests/sassim_t.c:573]: Removed.

**Review by rzinsly — APPROVED** [review]
Hi @munroesj52 sorry for the delay, the code looks good to me with some small clean-up.
I'm pushing this with the following changes.

**Pull request closed without merging**